packages/flutter/lib/src/material/mergeable_material.dart

```diff
@@ -0,0 +1,605 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+
+import 'dart:ui' show lerpDouble;
+
+/// A class that abstracts over a Material slice and a Material gap.
+///
+/// All [MergeableMaterialItem] objects need a [LocalKey].
+abstract class MergeableMaterialItem {
+  MergeableMaterialItem(this.key) {
+    assert(key != null);
+  }
+
+  final LocalKey key;
+}
+
+/// A class that can be used as a child to [MergeableMaterial]. It is a slice
+/// of [Material] that animates merging with other slices.
+///
+/// All [MaterialSlice] objects need a [LocalKey].
+class MaterialSlice extends MergeableMaterialItem {
+  /// Creates a slice of [Material] that's mergeable within a
+  /// [MergeableMaterial].
+  MaterialSlice({
+    LocalKey key,
```

Comment:
Here and elsewhere: we use @required (along with the assert you've already included) for required paramters.